libs/libarchfpga/src/read_fpga_interchange_arch.cpp

```diff
@@ -1284,126 +1479,156 @@ struct ArchReader {
             port->num_pins = 1;
             port->type = dir;
             port->is_clock = false;
 
-            bool is_input = dir == IN_PORT;
+            bool is_input = (dir == IN_PORT);
             port->name = is_input ? vtr::strdup(ipin.c_str()) : vtr::strdup(opin.c_str());
             port->model_port = nullptr;
             port->port_class = vtr::strdup(nullptr);
             port->port_power = (t_port_power*)vtr::calloc(1, sizeof(t_port_power));
+
+            port_idx++;
         }
 
         // OPAD mode
-        auto omode = &pad->modes[0];
-        omode->name = vtr::strdup("opad");
-        omode->parent_pb_type = pad;
-        omode->index = 0;
-        omode->num_pb_type_children = 1;
-        omode->pb_type_children = new t_pb_type[1];
-
-        auto opad = new t_pb_type;
-        opad->name = vtr::strdup("opad");
-        opad->num_pb = 1;
-        opad->parent_mode = omode;
-
-        num_ports = 1;
-        opad->num_ports = num_ports;
-        opad->ports = (t_port*)vtr::calloc(num_ports, sizeof(t_port));
-        opad->blif_model = vtr::strdup(MODEL_OUTPUT);
-        opad->model = get_model(arch_, std::string(MODEL_OUTPUT));
-
-        opad->ports[0] = get_generic_port(arch_, opad, IN_PORT, "outpad", MODEL_OUTPUT);
-        omode->pb_type_children[0] = *opad;
+        if (is_out) {
+            auto* omode = &pad->modes[mode_idx];
+            omode->name = vtr::strdup("opad");
+            omode->parent_pb_type = pad;
+            omode->index = mode_idx;
+
+            omode->num_pb_type_children = 1;
+            omode->pb_type_children = new t_pb_type[1];
+            auto* opad = &omode->pb_type_children[0];
+            opad->name = vtr::strdup("opad");
+            opad->num_pb = 1;
+            opad->parent_mode = omode;
+
+            num_ports = 1;
+            opad->num_ports = num_ports;
+            opad->ports = (t_port*)vtr::calloc(num_ports, sizeof(t_port));
+            opad->blif_model = vtr::strdup(MODEL_OUTPUT);
+            opad->model = get_model(arch_, std::string(MODEL_OUTPUT));
+
+            opad->ports[0] = get_generic_port(arch_, opad, IN_PORT, "outpad", MODEL_OUTPUT);
+
+            omode->num_interconnect = 1;
+            omode->interconnect = new t_interconnect[1];
+            auto* o_ic = &omode->interconnect[0];
+
+            std::string opad_istr = std::string(pad->name)  + std::string(".") + ipin;
+            std::string opad_ostr = std::string(opad->name) + std::string(".outpad");
+            std::string o_ic_name = std::string(pad->name)  + std::string("_") + std::string(opad->name);
+
+            o_ic->name = vtr::strdup(o_ic_name.c_str());
+            o_ic->type = DIRECT_INTERC;
+            o_ic->parent_mode_index = mode_idx;
+            o_ic->parent_mode = omode;
+            o_ic->input_string = vtr::strdup(opad_istr.c_str());
+            o_ic->output_string = vtr::strdup(opad_ostr.c_str());
+
+            mode_idx++;
+        }
```

Comment:
I wonder if this can actually be generalized somehow, so that we do not repeat code for both `inp` and `out`

Reply:
Yeah, so I haven't checked that on 7-series yet. I'll do that.